src/mlpack/methods/fastmks/fastmks_impl.hpp

```diff
@@ -185,10 +223,60 @@ void FastMKS<KernelType, MatType, TreeType>::Train(Tree* tree)
   this->metric = metric::IPMetric<KernelType>(tree->Metric().Kernel());
   this->setOwner = false;
 
-  if (treeOwner && referenceTree)
-    delete referenceTree;
+  if (treeOwner && this->referenceTree)
+    delete this->referenceTree;
 
   this->referenceTree = tree;
+  this->treeOwner = false;
+}
+
+template<typename KernelType,
+         typename MatType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType>
+void FastMKS<KernelType, MatType, TreeType>::Train(Tree& tree)
+{
+  if (naive)
+    throw std::invalid_argument("cannot call FastMKS::Train() with a tree when "
+        "in naive search mode");
+
+  if (setOwner)
+    delete this->referenceSet;
+
+  this->referenceSet = &this->tree->Dataset();
```

Comment:
What is `&this->tree`?  This probably does not compile... consider updating `Train(Tree*)` to simply call this method so that when you test `Train(Tree*)`, this is called too.  Also, the heavy use of `this->` in this function is not necessary.